### PR TITLE
`SHOW METRICS INFO` via query

### PR DESCRIPTION
**Pull request by Josipmrden**
New Enterprise query `SHOW METRICS INFO` has been added. It exposes all the metrics available through the HTTP metrics port (default 9091).

### CODE REVIEW

**Review by antepusic — APPROVED** [review]
Approving this in line with my previous comment

File: src/query/interpreter.cpp [hunk 1]
```diff
@@ -3954,6 +3954,55 @@ PreparedQuery PrepareDatabaseInfoQuery(ParsedQuery parsed_query, bool in_explici
...
+        throw QueryRuntimeException("SHOW METRICS INFO command is only available with a valid Enterprise License!");
+      }
+#else
+      throw QueryRuntimeException("SHOW METRICS INFO command is only available in Memgraph Enterprise build!");
+#endif
+      header = {"name", "type", "metric type", "value"};
+      handler = [storage = current_db.db_acc_->get()->storage()] {
+        auto info = storage->GetBaseInfo();
+        auto metrics_info = storage->GetMetrics();
+        std::vector<std::vector<TypedValue>> results;
```
Comment (as51340) [inline src/query/interpreter.cpp:3971]: All this code is good, to make it better:
- info and metrics_info can be const
- use std::ranges::sort

File: src/storage/v2/storage.cpp [hunk 2]
```diff
@@ -88,6 +90,34 @@ Storage::Accessor::Accessor(Accessor &&other) noexcept
 
 StorageMode Storage::GetStorageMode() const noexcept { return storage_mode_; }
 
+std::vector<EventInfo> Storage::GetMetrics() noexcept {
+  std::vector<EventInfo> result;
+  auto info = this->GetBaseInfo();
+
+  for (auto i = 0; i < memgraph::metrics::CounterEnd(); i++) {
+    result.emplace_back(memgraph::metrics::GetCounterName(i), memgraph::metrics::GetCounterType(i), "Counter",
```
Comment (as51340) [inline src/storage/v2/storage.cpp:98]: Use some variable name for Counter instead of black magic

File: src/storage/v2/storage.cpp [hunk 3]
```diff
@@ -88,6 +90,34 @@ Storage::Accessor::Accessor(Accessor &&other) noexcept
...
+                        memgraph::metrics::global_counters[i].load(std::memory_order_acquire));
```
Comment (as51340) [inline src/storage/v2/storage.cpp:99]: no need to use memgraph::metrics, use just metrics::
and I think you don't need to use load, without load should suffice

File: src/storage/v2/storage.cpp [hunk 4]
```diff
@@ -88,6 +90,34 @@ Storage::Accessor::Accessor(Accessor &&other) noexcept
...
+  }
+
+  for (auto i = 0; i < memgraph::metrics::GaugeEnd(); i++) {
+    result.emplace_back(memgraph::metrics::GetGaugeName(i), memgraph::metrics::GetGaugeTypeString(i), "Gauge",
+                        memgraph::metrics::global_gauges[i].load(std::memory_order_acquire));
```
Comment (as51340) [inline src/storage/v2/storage.cpp:104]: Similar comments

File: src/storage/v2/storage.cpp [hunk 5]
```diff
@@ -88,6 +90,34 @@ Storage::Accessor::Accessor(Accessor &&other) noexcept
...
+  }
+
+  for (auto i = 0; i < memgraph::metrics::HistogramEnd(); i++) {
+    const auto *name = memgraph::metrics::GetHistogramName(i);
+    auto &histogram = memgraph::metrics::global_histograms[i];
+
+    for (auto &[percentile, value] : histogram.YieldPercentiles()) {
+      auto metric_name = std::string(name) + "_" + std::to_string(percentile) + "p";
+
```
Comment (as51340) [inline src/storage/v2/storage.cpp:113]: use fmt::format

File: src/storage/v2/storage.cpp [hunk 6]
```diff
@@ -88,6 +90,34 @@ Storage::Accessor::Accessor(Accessor &&other) noexcept
...
+      result.emplace_back(metric_name, memgraph::metrics::GetHistogramType(i), "Histogram", value);
```
Comment (as51340) [inline src/storage/v2/storage.cpp:114]: std::move metric_name

Comment (as51340) [reply src/storage/v2/storage.cpp:114]: constant for histogram

File: src/storage/v2/storage.hpp [hunk 7]
```diff
@@ -420,6 +428,8 @@ class Storage {
 
   virtual StorageInfo GetBaseInfo() = 0;
 
+  virtual std::vector<EventInfo> GetMetrics() noexcept;
```
Comment (as51340) [inline src/storage/v2/storage.hpp:431]: does it need to be virtual when both storage modes use same impl?

**Review by as51340 — CHANGES_REQUESTED** [review]
Nice work, have few comments, nothing too important :)

**Comment by as51340** [comment]
maybe in general, consider rebasing over master instead of merging master into branch, history is a bit cleaner then

File: src/query/interpreter.cpp [hunk 8]
```diff
@@ -3962,6 +3962,55 @@ PreparedQuery PrepareDatabaseInfoQuery(ParsedQuery parsed_query, bool in_explici
         return std::pair{results, QueryHandlerResult::COMMIT};
       };
 
+      break;
+    }
+    case DatabaseInfoQuery::InfoType::METRICS: {
```
Comment (as51340) [inline src/query/interpreter.cpp:3967]: use emplace_back everywhere in this method

File: src/query/interpreter.cpp [hunk 9]
```diff
@@ -3962,6 +3962,55 @@ PreparedQuery PrepareDatabaseInfoQuery(ParsedQuery parsed_query, bool in_explici
...
+            {TypedValue("AverageDegree"), TypedValue("General"), TypedValue("Gauge"), TypedValue(info.average_degree)});
+        results.push_back({TypedValue("MemoryRes"), TypedValue("Memory"), TypedValue("Gauge"),
+                           TypedValue(static_cast<int64_t>(info.memory_res))});
+        results.push_back({TypedValue("DiskUsage"), TypedValue("Memory"), TypedValue("Gauge"),
+                           TypedValue(static_cast<int64_t>(info.disk_usage))});
+        for (const auto &metric : metrics_info) {
+          results.push_back({TypedValue(metric.name), TypedValue(metric.type), TypedValue(metric.event_type),
+                             TypedValue(static_cast<int64_t>(metric.value))});
+        }
+        std::ranges::sort(results.begin(), results.end(), [](const auto &record_1, const auto &record_2) {
```
Comment (as51340) [inline src/query/interpreter.cpp:3994]:
```suggestion
        std::ranges::sort(results, [](const auto &record_1, const auto &record_2) {
```

Comment (as51340) [reply src/query/interpreter.cpp:3994]: your lambda comparator can be simpler:
```
auto const cmp = [](auto const &record1, auto const &record2) {
MG_ASSERT(record1[1].IsString()); // or some method similar to this exists
MG_ASSERT(record2[1].IsString()); // or some method similar to this exists

return record1[1].ValueString().compare(record2[1].ValueString());
```
};

File: src/storage/v2/disk/storage.cpp [hunk 10]
```diff
@@ -1861,6 +1871,7 @@ void DiskStorage::DiskAccessor::UpdateObjectsCountOnAbort() {
   auto *disk_storage = static_cast<DiskStorage *>(storage_);
   uint64_t transaction_id = transaction_.transaction_id;
 
+  auto delta_size = transaction_.deltas.size();
```
Comment (as51340) [inline src/storage/v2/disk/storage.cpp:1874]: define it closer to where you use it

File: src/storage/v2/inmemory/storage.cpp [hunk 11]
```diff
@@ -1021,6 +1027,7 @@ void InMemoryStorage::InMemoryAccessor::GCRapidDeltaCleanup(std::list<Gid> &curr
   auto *mem_storage = static_cast<InMemoryStorage *>(storage_);
 
   auto const unlink_remove_clear = [&](std::deque<Delta> &deltas) {
+    auto delta_size = deltas.size();
```
Comment (as51340) [inline src/storage/v2/inmemory/storage.cpp:1030]: define it closer to where you use it.

File: src/storage/v2/inmemory/storage.cpp [hunk 12]
```diff
@@ -1958,7 +1969,14 @@ void InMemoryStorage::CollectGarbage(std::unique_lock<utils::ResourceLock> main_
       guard.unlock();
       // if lucky, there are no active transactions, hence nothing looking at the deltas
       // remove them now
+
+      size_t total_delta_size = 0;
+      for (const auto &gc_deltas : unlinked_undo_buffers) {
+        total_delta_size += gc_deltas.deltas_.size();
+      }
```
Comment (as51340) [inline src/storage/v2/inmemory/storage.cpp:1976]:
```suggestion
     auto const sum_func = [](uint32_t curr, auto const &gc_deltas) {
       return curr + gc_deltas.deltas_.size();
      }

      uint32_t total_delta_size = std::accumulate(unlinked_undo_buffers.begin(), unlinked_undo_buffers.end(), 0U, sum_func);
    
```
This should be correct, maybe I have some typo

File: src/storage/v2/mvcc.hpp [hunk 13]
```diff
@@ -123,6 +132,9 @@ inline Delta *CreateDeleteObjectDelta(Transaction *transaction, std::list<Delta>
     return nullptr;
   }
   transaction->EnsureCommitTimestampExists();
+  memgraph::utils::OnScopeExit increment_unreleased_deltas(
```
Comment (as51340) [inline src/storage/v2/mvcc.hpp:135]: You can here increment it without OnScopeExit, not sure what do you get?

File: src/storage/v2/storage.cpp [hunk 14]
```diff
@@ -88,6 +90,35 @@ Storage::Accessor::Accessor(Accessor &&other) noexcept
...
+    result.emplace_back(metrics::GetGaugeName(i), metrics::GetGaugeTypeString(i), kGaugeName,
+                        metrics::global_gauges[i]);
+  }
+
+  for (auto i = 0; i < metrics::HistogramEnd(); i++) {
+    const auto *name = metrics::GetHistogramName(i);
+    auto const &histogram = metrics::global_histograms[i];
+
+    for (auto &[percentile, value] : histogram.YieldPercentiles()) {
+      auto metric_name = fmt::format("{0}_{1}p", std::string(name), std::to_string(percentile));
```
Comment (as51340) [inline src/storage/v2/storage.cpp:114]: I think fmt can accept const char *, no need to convert it to string

**Review by as51340 — APPROVED** [review]
Few small changes required from the code side but you have approve

**Comment by Josipmrden** [comment]
Please delete either the [master < EPIC] or [master < Task] part, depending on what are your needs.

[master < Task] PR
- [X] Provide the full content or a guide for the final git message
    - **Added `SHOW METRICS INFO` query that exposes the system metrics via query**


### CI Testing Labels
Please select the appropriate CI test labels _(CI -build=**build-name** -test=**test-suite**)_


### Documentation checklist
- [X] Add the documentation label tag
- [X] Add the bug / feature label tag
- [X] Add the milestone for which this feature is intended
- [X] Write a release note, including added/changed clauses
    - **Added a new Enterprise query, `SHOW METRICS INFO`, that exposes the system metrics via query (same ones that are exposed through the HTTP monitoring endpoint at default port 9091)**
- [X] Link the documentation PR here
    - **https://github.com/memgraph/documentation/pull/809**
- [X] Tag someone from docs team in the comments @kgolubic

**Comment by kgolubic** [comment]
For release note I'll use:

- Added a new query, `SHOW METRICS INFO`, which exposes the same metrics
available from the HTTP endpoint, now accessible via a query.
[#2052](https://github.com/memgraph/memgraph/pull/2052)

**Pull request merged**